packages/headless-react/src/search-engine.tsx

```diff
@@ -0,0 +1,108 @@
+import {Controller, CoreEngine, SearchEngine} from '@coveo/headless';
+import {
+  ControllerDefinitionsMap,
+  InferControllerFromDefinition,
+  SearchEngineDefinitionOptions,
+  defineSearchEngine as defineBaseSearchEngine,
+  mapObject,
+} from '@coveo/headless/ssr';
+import {
+  Context,
+  createContext,
+  useContext,
+  useCallback,
+  useMemo,
+  useSyncExternalStore,
+} from 'react';
+import {
+  ContextCSRState,
+  ContextState,
+  ControllerHook,
+  InferControllerHooksMapFromDefinition,
+  ReactSearchEngineDefinition,
+} from './types';
+
+export function capitalize(s: string) {
+  return `${s.slice(0, 1).toUpperCase()}${s.slice(1)}`;
+}
+
+function isCSRContext<
+  TEngine extends CoreEngine,
+  TControllers extends ControllerDefinitionsMap<TEngine, Controller>,
+>(
+  ctx: Exclude<ContextState<TEngine, TControllers>, null>
+): ctx is ContextCSRState<TEngine, TControllers> {
+  return 'engine' in ctx;
+}
+
+function buildControllerHook<
+  TEngine extends CoreEngine,
+  TControllers extends ControllerDefinitionsMap<TEngine, Controller>,
+  TKey extends keyof TControllers,
+>(
+  context: Context<ContextState<TEngine, TControllers>>,
+  key: TKey
+): ControllerHook<InferControllerFromDefinition<TControllers[TKey]>> {
+  return () => {
+    const ctx = useContext(context);
+    if (ctx === null) {
+      throw 'Missing SSR state or CSR provider.';
+    }
+    const subscribe = useCallback(
+      (listener: () => void) =>
+        isCSRContext(ctx) ? ctx.controllers[key].subscribe(listener) : () => {},
+      [ctx]
+    );
+    const getSSRState = useCallback(() => ctx.controllers[key].state, [ctx]);
+    // TODO: Eval need for `useSyncMemoizedStore() instead of useSyncExternalStore()`
```

Comment:
Should these be linked to some investigation JIRA ? 
If only to explain why these alternative would be better ?

Reply:
+1

I think most (if not all) TODOs should be linked to a JIRA as to not fall through the cracks